### ISSUE TITLE
Commenting on Unreal Client or Server functions

### ISSUE THREAD

**Issue by KungFuDonkey** [doxygen/doxygen#9527]
**Describe the bug**
So this is not really a bug, but for a lack of having a community page I did not know where to post this, so I will post it here. I am working in unreal and I am using doxygen comments to generate a documentation website. The game is a multiplayer game and is thus using server and client functions that are created by the unreal header tool. The names of the server and client functions differ from the standard name. E.G.

ExampleClass.h
```
/// @brief class doxygen comment
UCLASS()
class AExampleClass : AActor
{
    GENERATED_BODY()
public:
    UFUNCTION(Client, Reliable);
    ExampleFunction();
}
```
ExampleClass.cpp
```
/// @brief function doxygen comment
ExampleFunction_Implementation()
{
}
```
The web page that is generated shows the class and the ExampleFunction(), but not the comment placed in the cpp file (neither brief nor detailed). I cannot change the name of ExampleFunction_Implementation() as this is used by unreal build tool internally to execute server and client functions. This is of course expected as the names do not match but, I was wondering if there is a way to do this, or should I move the comment to the header file instead? If this is not possible could this be added to a future release of doxygen?

**Expected behavior**
The web page either shows the ExampleFunction_implementation() documentation on the webpage, either under ExampleFunction() or ExampleDocumentation();

**Screenshots**
![image](https://user-images.githubusercontent.com/43845930/185431928-d14a3932-7573-46ad-a8df-5cb47413eab6.png)

![image](https://user-images.githubusercontent.com/43845930/185432532-607627b4-58b4-4091-8e72-833560fd000b.png)

![image](https://user-images.githubusercontent.com/43845930/185432435-852fb382-e261-4005-bb7c-e9cf86dca719.png)

![image](https://user-images.githubusercontent.com/43845930/185431826-50867e88-e0e6-4e31-b0bc-3d6772b04ca8.png)

expected:

![image](https://user-images.githubusercontent.com/43845930/185433114-67a5d151-e742-45d2-b87e-5e7d47a62eb3.png)

**To Reproduce**
Create an Unreal C++ project, put create an actor, create a server or client function, put a comment above the implementation

**Version**
1.9.4 

**Stack trace**
Not needed as doxygen did not crash

**Additional context**
--

**Comment by albert-github** [comment]
- Important to know is also what is in your doxygen settings file (i.e. the difference between the standard setting and the used settings, use `doxygen -x Doxyfile`).
- did you try to use `EXTRACT_ALL = YES`

**Comment by KungFuDonkey** [comment]
Yes even with `EXTRACT_ALL = YES` the result stays the same. The only difference I get is that I do get an empty detailed description, but there is still no _Implementation methods;
![image](https://user-images.githubusercontent.com/43845930/185533561-4d532935-7bea-4fb8-9a94-c17f934ae514.png)
![image](https://user-images.githubusercontent.com/43845930/185533813-5fe7ee45-f780-4a3f-9b44-5c9acfaa7441.png)
![image](https://user-images.githubusercontent.com/43845930/185533890-7eeaf5ed-11e1-469a-b010-2ee5b0643c42.png)

**Comment by albert-github** [comment]
- Still important to know is also what is in your doxygen settings file (i.e. the difference between the standard setting and the used settings, use `doxygen -x Doxyfile`).
- I cannot relate the images in the question to the given code snippet (missing settings, include files, ....):

So all in all:
- Can you please attach a, small, self contained example (source+configuration file in a, compressed, tar or zip file) that allows us to reproduce the problem? Please don't add external links as they might not be persistent.

**Comment by KungFuDonkey** [comment]
Of course I can do that, for the sake of it being small I will only include an example and some of the source files. Note that I generated the doxygen with the doxywizard (I do not know if that matters). Hope this helps!

[ExampleForDoxygen.zip](https://github.com/doxygen/doxygen/files/9381121/ExampleForDoxygen.zip)

**Comment by albert-github** [comment]
- doxywizard shouldn't won't make a difference
- where are e.g. `GENERATED_BODY()`, `YOMAIRDROP_API`, `UFUNCTION(...)` defined? Doxygen has no knowledge of these. And they are no probably skipped as the default setting is `SKIP_FUNCTION_MACROS = YES`
- Can you indicate what is missing in the example?

**Comment by KungFuDonkey** [comment]
These macros are defined by Unreal Engine and are needed for the Unreal Header tool, they should not really matter but I can set that setting to yes if needed. 

What is missing? If you run doxygen it generates the html website and if you go to the class_u_example_object.html page you can see that only the normal function has documentation and the others do not.

**Comment by KungFuDonkey** [comment]
After looking through my doxygen file, it already specifies that SKIP_FUNCITON_MACROS is set to YES

**Comment by albert-github** [comment]
But how can doxygen make a connection between e.g. `void UExampleObject::BlueprintNativeFunction_Implementation() {}` and `    UFUNCTION(BlueprintNativeEvent)
    void BlueprintNativeFunction();`?

You will probably also have noticed the warnings:
```
.../ExampleForDoxygen/source/Private/ExampleObject.cpp:12: warning: documented symbol 'void UExampleObject::ServerFunction_Implementation' was not declared or defined.
.../ExampleForDoxygen/source/Private/ExampleObject.cpp:16: warning: documented symbol 'bool UExampleObject::ServerFunction_Validate' was not declared or defined.
.../ExampleForDoxygen/source/Private/ExampleObject.cpp:19: warning: documented symbol 'void UExampleObject::ClientFunction_Implementation' was not declared or defined.
.../ExampleForDoxygen/source/Private/ExampleObject.cpp:22: warning: documented symbol 'void UExampleObject::BlueprintNativeFunction_Implementation' was not declared or defined.
```


The `SKIP_FUNCTION_MACROS = YES` is indeed the default settings but will result in that in e.g. the setting:
```
UFUNCTION(BlueprintNativeEvent)
    void BlueprintNativeFunction();
```
the `UFUNCTION(BlueprintNativeEvent)` will be removed, so doxygen will see:
```
    void BlueprintNativeFunction();
```

**Comment by KungFuDonkey** [comment]
Okay in what doxygen cpp file can I edit this so that it will find the _Implementation and make a connection with that. I cannot remove the _Implementation from the functions as these are needed for Unreal to build

**Comment by albert-github** [comment]
You could set:
```
MACRO_EXPANSION = YES
PREDEFINED = BlueprintNativeFunction_Implementation=BlueprintNativeFunction
```

**Comment by KungFuDonkey** [comment]
Look the thing is that the UFUNCTION() macro is NOT a C or C++ macro, it is a macro for the Unreal Header tool which generates a file with a lot of extra functions that includes the _Implementation (and _Validation). This file is then build with the CPP file. I do not want to use this file as it contains a lot of bloat from Unreal. 

The predefined version did not seem to work, but if it did it would not be very scalable with my project as I would have to set this for every function that is server, client or blueprint.

It would be better for me to modify the doxygen source code to allow the function make a link between these functions. If you could point me into the direction where I could do this I will make it myself. If you want I can turn it into a setting in the Doxyfile and create PR from it. This would then greatly improve compatibility with Unreal projects written in C++.

**Comment by albert-github** [comment]
Is there any (ISO-like) official standard regarding the "unreal" environment.
It think that manipulating doxygen for this is a bad idea and would probably require a lot of changes in a lot of files, probably a better option would be, as you probably have the Unreal Header tool result files to process these in one way or another (grep "_Implementation" ?) and make automatic defines out of these.

**Comment by KungFuDonkey** [comment]
To be honest I ran through the code of doxygen just now I think I will try to solve this myself and I will keep you updated, I currently think that it is possible by adding a couple of lines to doxygen.cpp line 6564 by adding an extra search if you are using a certain config tag and the function has _Implementation at the end of the function.

**Comment by KungFuDonkey** [comment]
So by adding an additional search like this:
![image](https://user-images.githubusercontent.com/43845930/185617852-425725ae-433f-4e5a-b064-4913cb40f4bb.png)

I am able to get the documentation on the html:
![image](https://user-images.githubusercontent.com/43845930/185617899-3195c9ad-f44d-443e-8bb4-0cc10b47b7e7.png)

**Comment by KungFuDonkey** [comment]
I did notice that the _Implementation is still showing in the detailed description, but that is fine for us I think as it shows where the documentation came from, but I don't know if this is bad for the rest of the program.

**Comment by albert-github** [comment]
Some remarks:
- Is there any (ISO-like) official standard regarding the "unreal" environment.
- don't use images to communicate changes, when someone would like to test one has to type all the code again instead of an easy cut / past and it would also be error-prone.
- would this be the only place? (how about searches in the output, caller / call graphs ...)
- this is of course very specific for the unreal situation
- having fixed names (in earlier comments you spoke about server names and client name, how do they differ)  in the code would not be a good idea, when doing it one should place the extra search(es) in the configuration variable
- You wrote "I did notice that the _Implementation is still showing in the detailed description, but that is fine for us" but that means it will be a hack and less good for doxygen.

**Comment by KungFuDonkey** [comment]
Is there any (ISO-like) official standard regarding the "unreal" environment.
- No there is no ISO-like standard as Unreal is a game engine, but I unreal uses _Implementation for all of the UFUNCTIONS that need some extra flexibility (like the server function needs a Validate and a Implementation version in order to be compiled)

don't use images to communicate changes, when someone would like to test one has to type all the code again instead of an easy cut / past and it would also be error-prone.
- Sorry I will use code snippets from now on

would this be the only place? (how about searches in the output, caller / call graphs ...)
- Yes, so far I have not encountered another function where this is needed but I will keep investigating

this is of course very specific for the unreal situation
- Yes, this is why I suggested to make a setting out of it and adding that to the if statement, idk yet how to add a config param right now
`
if (mn==0 && ALLOW_UNREAL_SEARCHES)
{
  QCString extraSearch = substitute(funcName, "_Implementation", "");
  mn = Doxygen::memberNameLinkedMap->find(extraSearch);
}
`

having fixed names (in earlier comments you spoke about server names and client name, how do they differ) in the code would not be a good idea, when doing it one should place the extra search(es) in the configuration variable
- I do not know exactly what you mean with this, but like I stated before every UFUNCTION that unreal deems "needs more flexibility" get an _Implementation added to their name in the CPP file

You wrote "I did notice that the _Implementation is still showing in the detailed description, but that is fine for us" but that means it will be a hack and less good for doxygen.
- This I agree with, but for that I will need to have extra information about how these comments are generated to change this in my PR

**Comment by KungFuDonkey** [comment]
Otherwise if you do not like the hard coded _Implementation it is of course possible to make an extra search list to do this and allow a user to define their own extensions for this, may someone need that for a different use case

**Comment by albert-github** [comment]
https://github.com/doxygen/doxygen/issues/9527#issuecomment-1220658199 this is exactly what I meant.

**Comment by KungFuDonkey** [comment]
Well I can try to make a PR this weekend or do you still suggest to not do it?

**Comment by albert-github** [comment]
I personally think it should not be in the doxygen code (seen a.o. the possibilities of  other settings etc, like `PREDEFINED` and as far as I can see the limited audience), I don't know what @doxygen his opinion is about this.

**Comment by KungFuDonkey** [comment]
While we wait on @doxygen, I have made some proposed changes on my [fork](https://github.com/KungFuDonkey/doxygen/tree/master) if you want to check it out. I currently do have a solution in mind for the naming problem but it is kind of a hack, so I will wait to see if you have a better solution. The naming problem can be found in [doxygen.cpp](https://github.com/KungFuDonkey/doxygen/blob/master/src/doxygen.cpp) on line 6631. (I know you did not want me to send links but including a code snippet is to big for the change)

**Comment by doxygen** [comment]
I think a doxygen `INPUT_FILTER` that removes the `_Implementation` and `_Validate` suffixes might be a good solution here.

I did a quick check on the example, with this setting in the Doxyfile:

```doxyfile
INPUT_FILTER           = ./filter.sh
```

and this `filter.sh` script (it assumes a Unix environment, so for Windows you may need to adapt things):

```sh
#!/bin/sh
sed -E 's/_(Implementation|Validate)\(/\(/g' $@
```

So the script just looks for `_Implementation(` or `_Validate(` and replaces it by a `(`, it could be made more sophisticated if needed.

**Comment by KungFuDonkey** [comment]
Okay I will give that a try, thanks in advance!

**Comment by KungFuDonkey** [comment]
Okay that kind of worked, but only for classes that actually have an _Implementation, the other ones are removed from the documentation entirely.

The command that I am running is:  `sed -E "s/_Implementation\\(/\(/g" %1` which gives me the expected output for a file with _implementation

**Comment by KungFuDonkey** [comment]
I think it is fixed. I added `.cpp=filter.bat` to FILTER_PATTERNS instead which does not remove all of the .h files from my documentation. Thanks @doxygen for this solution!

**Comment by doxygen** [comment]
Ok good to hear, I'll mark this issue as 'usage' as no changes to doxygen itself are required. It will be closed automatically when the next version of doxygen is released.

**Comment by KungFuDonkey** [comment]
But there is one more question on my mind, why are the other files removed? is there a way that I can fix that?

**Comment by albert-github** [comment]
Most likely something in the `filter.bat` resulting in an empty file.
What happens when you run outside doxygen th filter.bat on the missing file (say the file `missing.cpp`), so `filter.bat missing.bat` ?

**Comment by KungFuDonkey** [comment]
I get the same file without any changes, which would indicate that nothing should happen in doxygen?

**Comment by doxygen** [comment]
You can run doxygen with the `-d filteroutput` option to have it output the results after filtering. Maybe that gives a clue what is wrong.

**Comment by albert-github** [comment]
The output as written to stdout from the filter process is used by doxygen.
Please create small example (probably the `filter.bat`, `Doxyfile`, a correct file and a wrong file would be enough).

**Comment by KungFuDonkey** [comment]
-d filteroutput just shows the standard files, but alas they are not added to the documentation. I will create another example after my meeting (~1 hour) and I will send a zip here

**Comment by albert-github** [comment]
Good that you found the problem.
- adding a remark to the documentation is a good suggestion.
- that it sometimes worked and not in other times is probably sheer luck, for me no example needed.


(PS. nice short meeting)

**Comment by KungFuDonkey** [comment]
Jup nice and short meeting :D. Thanks for the amazing project and the help on this issue! It is really appreciated <3

**Comment by albert-github** [comment]
Regarding the documentation I've just pushed a proposed patch, pull request #9533

**Comment by albert-github** [comment]
Documentation has been integrated in master on GitHub (please don't close the issue as this will be done at the moment of an official release).

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.9.5.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
